### PR TITLE
add NVML binding

### PR DESCRIPTION
**Pull request by PatWie**
write NVML wrapper inspired by py-nvml but written from scratch.

### CODE REVIEW

**Comment by PatWie** [comment]
Not sure, if you want to use a separate file `nvml.py`. Here is a stand-alone version:
https://gist.github.com/PatWie/68abc24de854ba87ff07c68dea61e778

**Comment by ppwwyyxx** [comment]
A separate and __simpler__ file would be much better.

**Comment by PatWie** [comment]
I think

```python
    nvidia = NvidiaContext()
    nvidia.create_context()

    num_gpus = nvidia.NumCudaDevices()

    for device in nvidia.Devices():
        print(device.Name())
        print(device.UUID())

        print(device.Memory())
        print(device.Utilization())

    nvidia.destroy_context()
```

is pretty simple. For proper usage of ctypes you need all these exception handling to not left the user with unkown errors.

File: tensorpack/utils/nvml.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# File: nvml.py
+
+from ctypes import (byref, c_uint, c_ulonglong,
+                    CDLL, create_string_buffer, NVML_ERROR_FUNCTION_NOT_FOUND,
+                    NVML_ERROR_LIBRARY_NOT_FOUND, POINTER, Structure)
```
Comment (ppwwyyxx) [inline tensorpack/utils/nvml.py]: This import line doesn't run.

File: tensorpack/utils/nvml.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+def CheckNvmlReturn(ret):
+    if (ret != 0):
+        raise NvmlException(ret)
+    return ret
+
+
+class NVML(object):
+
+    def load(self):
+        global nvmlLib
```
Comment (ppwwyyxx) [inline tensorpack/utils/nvml.py]: Not necessary to make it global, right?

File: tensorpack/utils/nvml.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+        Example:
+
+            >>> print(nvidia.Device(0).UUID())
+            GPU-xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+
+        Returns:
+            UUID of GPU
+        """
+        return _NVML.call(self.hnd, NVML_DEVICE_UUID_BUFFER_SIZE, "nvmlDeviceGetUUID").value
```
Comment (ppwwyyxx) [inline tensorpack/utils/nvml.py]: I don't think UUID and name of card will ever be useful. They can be removed together with call() and BUFFER constants.

File: tensorpack/utils/nvml.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        super(NvidiaContext, self).__init__()
+
+    def create_context(self):
+        """Create a new context
+        """
+        _NVML.load()
+        CheckNvmlReturn(_NVML.get_function("nvmlInit_v2")())
+
+        # Atomically update refcount
+        global nvml_init_count
```
Comment (ppwwyyxx) [inline tensorpack/utils/nvml.py]: Apart from +1 and -1 I don't see this variable used anywhere

File: tensorpack/utils/nvml.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        _lib_lock.acquire()
+        nvml_init_count += 1
+        _lib_lock.release()
+
+    def destroy_context(self):
```
Comment (ppwwyyxx) [inline tensorpack/utils/nvml.py]: Make it a with statement?
` with NVMLContext() as ctx: ...`

Comment (PatWie) [reply tensorpack/utils/nvml.py]: I thought about this as well. But for the callback version of your GPUUtilization call, you cannot use your way, as it closes and open the context for each epoch.

File: tensorpack/utils/nvml.py (hunk 2 above)
Comment (PatWie) [reply tensorpack/utils/nvml.py]: It think it is correct, I want a singleton-pattern here

Comment (ppwwyyxx) [reply tensorpack/utils/nvml.py]: But everytime load() is called the so file gets loaded again. 
What's singleton is `_NVML = NVML()`. This so can just be an attribute.

Comment (ppwwyyxx) [reply tensorpack/utils/nvml.py]: And even if it's made an attribute, is it correct to load the so everytime load() is called? Should it be cached?

**Pull request merged**